src/inline_table.rs

```diff
@@ -97,32 +97,26 @@ impl InlineTable {
     ///
     /// The comparison function receives two key and value pairs to compare (you can sort by keys or
     /// values or their combination as needed).
-    pub fn sort_values_by<F>(&mut self, compare: F)
+    pub fn sort_values_by<F>(&mut self, mut compare: F)
     where
-        F: Fn(&Key, &Item, &Key, &Item) -> std::cmp::Ordering,
+        F: FnMut(&Key, &Item, &Key, &Item) -> std::cmp::Ordering,
+    {
+        self.sort_values_by_internal(&mut compare);
+    }
+
+    fn sort_values_by_internal<F>(&mut self, compare: &mut F)
+    where
```

Comment:
Do we still need `_internal`?

Reply:
Yes. Because we want to recursively sort in a loop, we either need to

1. add `where F: Copy` as a restriction on the function, or
2. take a `&mut F` reference instead of taking ownership.

`F: Copy` seems too restrictive (like you mentioned previously, we want to get as close to `IndexMap::sort_by`'s requirements as possible: just an `FnMut`)

Changing the signature to `Table::sort_values_by<F>(&mut self, compare: &mut F)` is an option – and then we don't need `_internal`, but then the signature differs from the expectations that most people have from things like `Vec::sort_by` and `IndexMap::sort_by`.

And even though `&mut F` implements `FnMut`, we can't take ownership in `Table::sort_values_by(&mut self, mut compare: F)` and then make the recursive call with `table.sort_values_by(&mut compare)` because then Rust detects an infinite loop trying to do type inference.

So basically, the way to get it done is for the signature to be `Table::some_function(&mut self, compare: &mut F)` and the recursion to be `sub_table.some_function(compare /* pass along the &mut F */)`. `Table::sort_values_by_internal` fills that roll, and `Table::sort_values_by` provides a better interface to it.

Reply:
Ah, missed the subtle signature difference, thanks!